### ISSUE TITLE
Remote builders to survive ssh disconnect

### ISSUE THREAD

**Issue by volth** [NixOS/nix#1953]
It would be nice if remote builders could survive ssh disconnect.

**Comment by dezgeg** [comment]
Should be relatively easy to do if both sides have https://mistertea.github.io/EternalTerminal/.

**Comment by 7c6f434c** [comment]
@dezgeg I think it is better not to need a pty in the process. Also, an improvement in cancellation logic would be useful for a scenario where A and B depend on X, and A also depends on A1 while B depends on B1. If you build A and B from different terminals, A starts building both X and A1 but A1 fails, the build of X will get restarted for B instead of transferring the «ownership» without interrupting the process.

**Comment by dezgeg** [comment]
Despite what the name says, Eternal Terminal doesn't require a PTY, it can work in a dump-bidirectional-pipe-of-bytes mode.

**Comment by volth** [comment]
maybe running each build in a proper named session of ```screen``` would go.

**Comment by volth** [comment]
Perhaps it deserves its own issue, but there could be a single solution for both. Because both are about sandboxing the remote builder:

----
It would be nice to quote resources of remote builder.
Not by the server-side settings, but either client-side (sounds strange, but hey, even the max number of parallel bulders is a client-side setting from ```/etc/nix/machines```), or just by unconditionally imposing some sanity limits.

The main problem is the packages doing "fork bomb" (or other activity performed up until the limt ir reached) on the test phase.

**Comment by shlevy** [comment]
There are two parts to a solution here:

1. Don't die on `SIGHUP` in `nix-daemon --stdio` (for `ssh-ng://`) and `nix-store --serve` (for `ssh://`). This is trivial, and doesn't require anything like `screen` or something more complicated.
2. Stop listening for close events on the connection file descriptor, as is currently the case for at least `nix-daemon --stdio`. This should be gated by a specific request (possibly the default for ssh stores), as generally interrupt on disconnect (e.g. for local daemon connections) is desirable behavior.

**Comment by volth** [comment]
I meant to use ```screen``` to discover a running build upon reconnect, take ownership, and start ```tail -f```ing it.

**Comment by edolstra** [comment]
It seems pretty undesirable if builds continue to run when the requesting process has gone away. If I hit Ctrl-C on a nix-build, I expect the build to stop... Likewise, if I kill `hydra-queue-runner`, I expect all build machines to reach a quiescent state.

Also, even if the build continues to run, the client has no way to "resume" because it doesn't know about running builds. So it might start a new build on another machine.

**Comment by edolstra** [comment]
Maybe you can tweak your SSH/TCP keep-alive settings to prevent the connection from being killed when there are temporary network outages?

**Comment by dezgeg** [comment]
That doesn't help if the IP address changes for some reason, EternalTerminal would handle that gracefully . I don't know how it handles Ctrl-C though, but presumably patches could be (at leastattempted to be) sent upstream if it doesn't.

**Comment by shlevy** [comment]
IMO *if* we want to support this we should do it through a detach/reattach request in the protocol. That would also work for local daemon and other non-SSH scenarios, and doesn't require additional software.

**Comment by 7c6f434c** [comment]
I would say the requests would be «set the timeout for running in a detached state» (it is often impossible to notify that you have lost connection), «attach if detached» and «subscribe to a notification about availability of reattaching» (the latter would be useful in the common dependency takeover scenario).

**Comment by volth** [comment]
> If I hit Ctrl-C on a nix-build, I expect the build to stop... 

actually it does not stop on Ctrl-C or network disconnect (if the build was running on a remote slave)
it goes on abandoned.

**Comment by dtzWill** [comment]
@volth it should stop! I think at some point there was a bug where this didn't occur but AFAIK this has worked for quite some time (1.12-ish onwards). Please file an issue with details if you are seeing otherwise.

**Comment by edolstra** [comment]
There was a bug on macOS that prevented builds from stopping. It was fixed by @shlevy recently.

**Comment by dezgeg** [comment]
Where in the code it actually detects a network disconnect and/or Ctrl-C over SSH? I once had the need to have the remote `foo` process in `ssh user@host foo` reliably die on Ctrl-C and the only way I got that to work was to enable the KillUserProcesses for that user in systemd.

**Comment by shlevy** [comment]
@dezgeg https://github.com/NixOS/nix/blob/21ebf6a16028291914bd14b0e5463b020dc7affa/src/libutil/monitor-fd.hh

**Comment by dezgeg** [comment]
Hmm, on my machine I see `build-remote` processes still running even though the nix-store process that initiated that build is long gone.

**Comment by matthewbauer** [comment]
This seems like it could be really useful behind an option. Something like ```nix build --no-hup``` telling the Nix daemon to not stop the build when the SSH connection is disconnected would make remote builders a lot more useful for me. I end up staying at work for a while waiting for a long build to finish because the build was started on my laptop.
